src/probnum/linalg/solvers/_probabilistic_linear_solver.py

```diff
@@ -35,9 +60,279 @@ class ProbabilisticLinearSolver:
+    >>> from probnum.problems.zoo.linalg import random_spd_matrix
+
+    >>> rng = np.random.default_rng(42)
+    >>> n = 100
+    >>> A = random_spd_matrix(rng=rng, dim=n)
+    >>> b = rng.standard_normal(size=(n,))
+    >>> linsys = LinearSystem(A=A, b=b)
+
+    Create a custom probabilistic linear solver from pre-defined components.
+
+    >>> from probnum.linalg.solvers import (
+    ...     ProbabilisticLinearSolver,
+    ...     belief_updates,
+    ...     beliefs,
+    ...     information_ops,
+    ...     policies,
+    ...     stopping_criteria,
+    ... )
+
+    >>> pls = ProbabilisticLinearSolver(
+    ...     policy=policies.ConjugateGradientPolicy(),
+    ...     information_op=information_ops.ProjectedRHSInformationOp(),
+    ...     belief_update=belief_updates.solution_based.SolutionBasedProjectedRHSBeliefUpdate(),
+    ...     stopping_criterion=stopping_criteria.MaxIterationsStoppingCriterion(100)
+    ...     | stopping_criteria.ResidualNormStoppingCriterion(atol=1e-5, rtol=1e-5),
+    ... )
+
+    Define a prior over the solution.
+
+    >>> from probnum import linops, randvars
+    >>> prior = beliefs.LinearSystemBelief(
+    ...     x=randvars.Normal(
+    ...         mean=np.zeros((n,)),
+    ...         cov=np.eye(n),
+    ...     ),
+    ... )
+
+    Solve the linear system using the custom solver.
+
+    >>> belief, solver_state = pls.solve(prior=prior, problem=linsys)
+    >>> np.linalg.norm(linsys.A @ belief.x.mean - linsys.b) / np.linalg.norm(linsys.b)
+    7.1886e-06
+    """
+
+    def __init__(
+        self,
+        policy: policies.LinearSolverPolicy,
+        information_op: information_ops.LinearSolverInformationOp,
+        belief_update: belief_updates.LinearSolverBeliefUpdate,
+        stopping_criterion: stopping_criteria.LinearSolverStoppingCriterion,
+    ):
+        self.policy = policy
+        self.information_op = information_op
+        self.belief_update = belief_update
+        super().__init__(stopping_criterion=stopping_criterion)
+
+    def solve_iterator(
+        self,
+        prior: beliefs.LinearSystemBelief,
+        problem: problems.LinearSystem,
+        rng: Optional[np.random.Generator] = None,
+    ) -> Generator[LinearSolverState, None, None]:
+        """Generator implementing the solver iteration.
+
+        This function allows stepping through the solver iteration one step at a time
+        and exposes the internal solver state.
+
+        Parameters
+        ----------
+        prior
+            Prior belief about the quantities of interest :math:`(x, A, A^{-1}, b)` of the linear system.
+        problem
+            Linear system to be solved.
+        rng
+            Random number generator.
+
+        Yields
+        ------
+        solver_state
+            State of the probabilistic linear solver.
+        """
+        solver_state = LinearSolverState(problem=problem, prior=prior, rng=rng)
+
+        while True:
+
+            yield solver_state
+
```

Comment:
Since we pass the stopping criterion to the constructor and not the `solve` method, I think we should check the stopping criterion here and return if it is met. Otherwise the behavior of the iterator would seem somewhat unintuitive to me.